quic/s2n-quic-core/src/packet/interceptor.rs

```diff
@@ -95,6 +104,19 @@ where
     A: Interceptor,
     B: Interceptor,
 {
+    // Return the largest packet number given multiple composed Interceptor implementations.
+    #[inline(always)]
+    fn intercept_rx_inject_ack(&mut self, space: PacketNumberSpace) -> Option<VarInt> {
```

Comment:
This might have already been discussed, but is there a way this could be more generic, as in just `fn intercept_rx_frame(&mut self, frame: Frame) -> Frame` and you could just implement one that intercepts an Ack and modifies it?

Reply:
Tried this suggestion but ran into a type issue with the Frame<AckRanges, Data> type. I suspect we might also run into similar issues with the `Data` generic type.

The trait `s2n_quic_core::frame::ack::AckRanges` has different impls and it [mismatches](https://github.com/aws/s2n-quic/compare/ak-interceptFrame?expand=1#diff-ec45b781c59b2bd24edd9b8fa8f4bf79bd06c62bc780350e3c8f275f5d63ad0eR206-R208) in our return type.

Specifically, in `space/mod` the type is `s2n_quic_core::frame::ack::AckRangesDecoder`, while in the interceptor we construct a `s2n_quic_transport::ack::ack_ranges::AckRanges`

```
error[E0308]: mismatched types
   --> quic/s2n-quic/src/tests/skip_packets.rs:206:50
    |
182 |     fn intercept_rx_frame<'a, A: s2n_quic_core::frame::ack::AckRanges, D>(
    |                               - this type parameter
...
206 |                 s2n_quic_core::frame::Frame::Ack(ack)
    |                 -------------------------------- ^^^ expected type parameter `A`, found struct `s2n_quic_transport::ack::ack_ranges::AckRanges`
    |                 |
    |                 arguments to this enum variant are incorrect
    |
    = note: expected struct `s2n_quic_core::frame::Ack<A>`
               found struct `s2n_quic_core::frame::Ack<s2n_quic_transport::ack::ack_ranges::AckRanges>`
```

Reply:
I think I would prefer something like:

```rust
fn intercept_rx_ack<A: interceptor::Ack>(&mut self, ack: &mut A);
```

where 

```rust
pub trait Ack {
    fn space(&self) -> PacketNumberSpace;
    fn range(&mut self, range: RangeInclusive<VarInt>);
}
```

Reply:
I had a version working with the `Ack` trait, but I didn't see the purpose of it in the end. Let me know if my approach has any downsides compared to the trait

Reply:
My thinking with the trait was to avoid allocating for the ranges entirely and just dispatch directly to the `handle_ack_frame` function. Making it a trait means that the concrete space types don't leak into this interface.

Reply:
I'm not completely opposed to your approach. It just locks us into a concrete type for the range buffer.

Reply:
I thought the concrete `ack::Ranges` type was being kept internal to the `interceptor::Ack` type and not leaking into the interface... though is the `impl<'a> From<&'a Ack> for frame::Ack<&'a ack::Ranges>` the issue?

Reply:
The concrete type is your new `Ack` itself. So if that `Ack` struct started owning a reference to a space, it would need to be generic over some trait that dispatches ACK frames.

Reply:
Thanks, updated using the Trait. I had to add `Sized` to the `PacketSpace` trait, not sure if there is a downside to that